analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessary.cs

```diff
@@ -27,26 +27,48 @@ public sealed class BooleanLiteralUnnecessary : BooleanLiteralUnnecessaryBase<Bi
 
         protected override bool IsBooleanLiteral(SyntaxNode node) => IsTrueLiteralKind(node) || IsFalseLiteralKind(node);
 
-        protected override SyntaxNode GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
-
-        protected override SyntaxNode GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
-
-        protected override SyntaxToken GetOperatorToken(BinaryExpressionSyntax binaryExpression) => binaryExpression.OperatorToken;
+        protected override SyntaxToken? GetOperatorToken(SyntaxNode node) =>
+            node switch
+            {
+                BinaryExpressionSyntax binary => binary.OperatorToken,
+                _ when IsPatternExpressionSyntaxWrapper.IsInstance(node) => ((IsPatternExpressionSyntaxWrapper)node).IsKeyword,
+                { RawKind: (int)SyntaxKindEx.ConstantPattern } => GetOperatorToken(node.Parent),
+                { RawKind: (int)SyntaxKindEx.NotPattern } => GetOperatorToken(node.Parent),
+                _ => null,
+            };
 
         protected override bool IsTrueLiteralKind(SyntaxNode syntaxNode) => syntaxNode.IsKind(SyntaxKind.TrueLiteralExpression);
 
         protected override bool IsFalseLiteralKind(SyntaxNode syntaxNode) => syntaxNode.IsKind(SyntaxKind.FalseLiteralExpression);
 
-        protected override bool IsInsideTernaryWithThrowExpression(BinaryExpressionSyntax syntaxNode) =>
+        protected override bool IsInsideTernaryWithThrowExpression(SyntaxNode syntaxNode) =>
             syntaxNode.Parent is ConditionalExpressionSyntax conditionalExpression
             && (IsThrowExpression(conditionalExpression.WhenTrue) || IsThrowExpression(conditionalExpression.WhenFalse));
 
+        protected override SyntaxNode GetLeftNode(SyntaxNode node) =>
+            node switch
+            {
+                BinaryExpressionSyntax binaryExpression => binaryExpression.Left,
+                _ when IsPatternExpressionSyntaxWrapper.IsInstance(node) => ((IsPatternExpressionSyntaxWrapper)node).Expression,
+                _ => null
+            };
+
+        protected override SyntaxNode GetRightNode(SyntaxNode node) =>
+            node switch
+            {
+                BinaryExpressionSyntax binaryExpression => binaryExpression.Right,
+                _ when IsPatternExpressionSyntaxWrapper.IsInstance(node) => GetRightNode(((IsPatternExpressionSyntaxWrapper)node).Pattern),
+                { RawKind: (int)SyntaxKindEx.ConstantPattern } => ((ConstantPatternSyntaxWrapper)node).Expression,
+                { RawKind: (int)SyntaxKindEx.NotPattern } => GetRightNode(((UnaryPatternSyntaxWrapper)node).Pattern),
```

Comment:
This complicates a lot of logic: You return either an expression or a negated expression. I think we picked the wrong place for this. Can you try to return just the pattern
```cs
_ when IsPatternExpressionSyntaxWrapper.IsInstance(node) => (IsPatternExpressionSyntaxWrapper)node).Pattern,
```
and do the logic in `IsTrueLiteral` / `IsFalseLiteral`? By doing so, we could also add support for `b == !true` and the like (would return true for `IsFalseLiteral` for right).

Reply:
Can you try to remove the last two cases and return the pattern instead:
```cs
_ when IsPatternExpressionSyntaxWrapper.IsInstance(node) => (IsPatternExpressionSyntaxWrapper)node).Pattern,
```